### PR TITLE
Added hyperparameter tuning and solver performance

### PR DESCRIPTION
**Pull request by axiomcura**
Hello, this PR addresses a reviewer's request by adding the hyperparameter tuning scores, along with the corresponding solvers used during the `RandomizedSearchCV process`. These results are saved in the `./results/2.modeling` directory

Reviewer's comment: 
> Online Methods, Machine learning model training: It would be informative to include the hyperparameter search and solver evaluation results in the Supplementary Material.

### CODE REVIEW

**Review by gwaybio — APPROVED** [review]
I'm surprised there are so few rows - does this mean our hyperparameter search was this small?

**Comment by axiomcura** [comment]
Hmm, I'll double-check this. The hyperparameter search takes about 10 minutes to complete, so it's possible that I might be recording the results incorrectly.

I will notify you if there are any updates!

**Comment by axiomcura** [comment]
@gwaybio - I discovered why only 10 iterations are being shown: the `RandomizedSearchCV` is set to randomly test only 10 combinations from the `param_grid`. This is controlled by the `n_iter` parameter, which limits the search to a specified number of random configurations within `param_grid`. You can refer to the [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html) for more details.

```python
param_grid = {
    "penalty": ["l1", "l2", "elasticnet"],                        
    "C": [0.001, 0.01, 0.1, 1, 10, 100],                          
    "max_iter": np.arange(100, 1100, 100),                      
    "tol": np.arange(1e-6, 1e-3, 1e-6),                         
    "l1_ratio": [0.1, 0.3, 0.5, 0.7, 0.9],                        
    "solver": ["newton-cg", "lbfgs", "liblinear", "sag", "saga"],  # 5 options
}

# Execute RandomizedSearchCV
random_search = RandomizedSearchCV(
    lr_model,
    param_distributions=param_grid,
    n_iter=10,  # Randomly tests only 10 combinations from param_grid
    cv=5,
    verbose=0,
    random_state=0,
    n_jobs=-1,
)
```

If we were to perform a full grid search with the given parameters, it would require evaluating a total of:
```text
3 × 6 × 10 × 999 × 5 × 5 = 449,550 combinations
```
where each number represents the number of options available for each parameter.

Therefore, the `n_iter` parameter balances the trade-off between computational cost and search thoroughness. It enables us to explore a large hyperparameter space  without exhaustively testing every combination. As of right now, it takes around 10 minutes to trian the mode when `n_iter` = 10


Looking at the score with the best parameters, it scores 67%:
```text
2024-09-25 10:14:46,793 - INFO - Training non-shuffled-model
2024-09-25 10:14:46,794 - INFO - saving
2024-09-25 10:22:10,764 - INFO - All parameters tested during RandomizedSearchCV:
2024-09-25 10:22:10,765 - INFO - Iteration 1: {'tol': 0.000406, 'solver': 'saga', 'penalty': 'elasticnet', 'max_iter': 900, 'l1_ratio': 0.9, 'C': 0.01}
2024-09-25 10:22:10,765 - INFO - Iteration 2: {'tol': 0.00010499999999999999, 'solver': 'newton-cg', 'penalty': 'elasticnet', 'max_iter': 800, 'l1_ratio': 0.9, 'C': 0.1}
2024-09-25 10:22:10,765 - INFO - Iteration 3: {'tol': 0.000468, 'solver': 'saga', 'penalty': 'elasticnet', 'max_iter': 1000, 'l1_ratio': 0.9, 'C': 0.1}
2024-09-25 10:22:10,765 - INFO - Iteration 4: {'tol': 0.000732, 'solver': 'saga', 'penalty': 'elasticnet', 'max_iter': 800, 'l1_ratio': 0.9, 'C': 0.1}
2024-09-25 10:22:10,765 - INFO - Iteration 5: {'tol': 0.000565, 'solver': 'saga', 'penalty': 'elasticnet', 'max_iter': 500, 'l1_ratio': 0.5, 'C': 0.001}
2024-09-25 10:22:10,765 - INFO - Iteration 6: {'tol': 0.0009839999999999998, 'solver': 'liblinear', 'penalty': 'elasticnet', 'max_iter': 900, 'l1_ratio': 0.7, 'C': 1}
2024-09-25 10:22:10,765 - INFO - Iteration 7: {'tol': 0.000745, 'solver': 'newton-cg', 'penalty': 'elasticnet', 'max_iter': 100, 'l1_ratio': 0.5, 'C': 0.001}
2024-09-25 10:22:10,765 - INFO - Iteration 8: {'tol': 0.000313, 'solver': 'lbfgs', 'penalty': 'l1', 'max_iter': 200, 'l1_ratio': 0.5, 'C': 10}
2024-09-25 10:22:10,765 - INFO - Iteration 9: {'tol': 0.000627, 'solver': 'lbfgs', 'penalty': 'l1', 'max_iter': 100, 'l1_ratio': 0.7, 'C': 1}
2024-09-25 10:22:10,765 - INFO - Iteration 10: {'tol': 0.000887, 'solver': 'lbfgs', 'penalty': 'l1', 'max_iter': 400, 'l1_ratio': 0.1, 'C': 10}
2024-09-25 10:22:10,765 - INFO - Best parameters found: {'tol': 0.000468, 'solver': 'saga', 'penalty': 'elasticnet', 'max_iter': 1000, 'l1_ratio': 0.9, 'C': 0.1}
2024-09-25 10:22:10,765 - INFO - Best score: 0.6748749721493821
2024-09-25 10:22:11,224 - INFO - Training shuffled-model
```

**Comment by gwaybio** [comment]
Gotcha! 10 combinations (in 10 minutes) seems unnecessarily restrictive - if we increase to 100, do we see much higher than 67%?

**Comment by axiomcura** [comment]
@gwaybio - Just finished rerunning the hyperparameter search and the accuracy when from 67% to 75%

```text
2024-09-25 11:41:28,023 - INFO - Best parameters found: {'tol': 0.000651, 'solver': 'newton-cg', 'penalty': 'l2', 'max_iter': 1000, 'l1_ratio': 0.1, 'C': 10}
2024-09-25 11:41:28,023 - INFO - Best score: 0.7453232814026428
```

This took around 37 minutes to complete. Should I increase it to 150 or 200?

**Comment by axiomcura** [comment]
After increasing it to 100 iterations:

```text
2024-09-25 11:41:28,023 - INFO - Iteration 100: {'tol': 0.000713, 'solver': 'newton-cg', 'penalty': 'l2', 'max_iter': 900, 'l1_ratio': 0.9, 'C': 0.01}
2024-09-25 11:41:28,023 - INFO - Best parameters found: {'tol': 0.000651, 'solver': 'newton-cg', 'penalty': 'l2', 'max_iter': 1000, 'l1_ratio': 0.1, 'C': 10}
2024-09-25 11:41:28,023 - INFO - Best score: 0.7453232814026428
```


and 200 iterations
```text
2024-09-25 16:57:53,541 - INFO - Best parameters found: {'tol': 6.9e-05, 'solver': 'lbfgs', 'penalty': 'l2', 'max_iter': 600, 'l1_ratio': 0.9, 'C': 10}
2024-09-25 16:57:53,541 - INFO - Best score: 0.7474526111025417
2024-09-25 16:57:53,999 - INFO - Training shuffled-model
```


I will be merging this PR in order to get the main code base updated with everything else. We can continue our discussion here.
